File: model/src/main/scala/ru/mipt/acsl/decode/model/types/EnumType.scala
package ru.mipt.acsl.decode.model.types

import ru.mipt.acsl.decode.model.Referenceable
import ru.mipt.acsl.decode.model.naming.{Container, ElementName, Namespace}
import ru.mipt.acsl.decode.model.proxy.MaybeProxy

/**
  * @author Artem Shein
  */
trait EnumType extends DecodeType with Container {

  def isFinal: Boolean

  def constants: Seq[EnumConstant] = objects.flatMap {
    case c: EnumConstant => Seq(c)
    case _ => Seq.empty
  }

  def extendsTypeOption: Option[EnumType] = extendsOrBaseTypeProxy.left.toOption.map(_.obj)

  def baseTypeOption: Option[DecodeType] = extendsOrBaseTypeProxy.right.toOption.map(_.t)

  def extendsOrBaseTypeProxy: Either[MaybeProxy.Enum, TypeMeasure]

  def extendsOrBaseType: DecodeType = extendsOrBaseTypeProxy.fold(l => l.obj, r => r.t)

  def eitherExtendsOrBaseType: Either[EnumType, TypeMeasure] = extendsOrBaseTypeProxy.fold(l => Left(l.obj), r => Right(r))

  def baseType: DecodeType = eitherExtendsOrBaseType.fold(l => l.baseType, r => r.t)

  def systemName: String = "EnumType@" + hashCode()

  override def toString: String =
    s"${this.getClass}{alias = $alias, namespace = $namespace, extendsOrBaseTypeProxy = $extendsOrBaseTypeProxy," +
      s" objects = $objects, isFinal = $isFinal, typeParameters = $typeParameters}"

}

object EnumType {

  private class EnumTypeImpl(val alias: Option[Alias.NsType], var namespace: Namespace,
                             var extendsOrBaseTypeProxy: Either[MaybeProxy.Enum, TypeMeasure],
                             var objects: Seq[Referenceable], var isFinal: Boolean,
                             val typeParameters: Seq[ElementName])
    extends EnumType {

    def extendsTypeProxy: Option[MaybeProxy.Enum] = extendsOrBaseTypeProxy.left.toOption

    override def extendsTypeOption: Option[EnumType] = extendsTypeProxy.map(_.obj)

    def baseTypeProxy: MaybeProxy.TypeProxy =
      extendsOrBaseTypeProxy match {
        case Left(extendsType) => extendsType
        case Right(baseType) => baseType.typeProxy
      }

  }

  def apply(alias: Option[Alias.NsType], namespace: Namespace,
            extendsOrBaseTypeProxy: Either[MaybeProxy.Enum, TypeMeasure],
            objects: Seq[Referenceable], isFinal: Boolean,
            typeParameters: Seq[ElementName]): EnumType =
    new EnumTypeImpl(alias, namespace, extendsOrBaseTypeProxy, objects, isFinal, typeParameters)
}

File: model/src/main/java/ru/mipt/acsl/decode/model/types/EnumType.java
package ru.mipt.acsl.decode.model.types;

import org.jetbrains.annotations.Nullable;
import ru.mipt.acsl.decode.model.*;
import ru.mipt.acsl.decode.model.naming.Container;
import ru.mipt.acsl.decode.model.naming.ElementName;
import ru.mipt.acsl.decode.model.naming.Namespace;
import ru.mipt.acsl.decode.model.proxy.MaybeProxyCompanion;

import java.util.*;

/**
 * Created by metadeus on 07.06.16.
 */
public interface EnumType extends DecodeType, Container, HasNamespace {

    static EnumType newInstance(@Nullable Alias.NsType alias, Namespace namespace,
                                MaybeProxyEnumOrTypeMeasure extendsOrBaseTypeProxy,
                                List<Referenceable> objects, boolean isFinal,
                                List<ElementName> typeParameters) {
        return new EnumTypeImpl(alias, namespace, extendsOrBaseTypeProxy, objects, isFinal, typeParameters);
    }

    boolean isFinal();

    default List<EnumConstant> constants() {
        return filterByClass(EnumConstant.class);
    }

    default Optional<EnumType> extendsTypeOption() {
        return extendsOrBaseTypeProxy().maybeProxyEnum().map(MaybeProxyCompanion.Enum::obj);
    }

    default Optional<DecodeType> baseTypeOption() {
        return extendsOrBaseTypeProxy().typeMeasure().map(TypeMeasure::t);
    }

    default DecodeType baseType() { return baseTypeOption().orElseGet(() -> extendsTypeOption().get().baseType()); }

    MaybeProxyEnumOrTypeMeasure extendsOrBaseTypeProxy();

    default DecodeType extendsOrBaseType() {
        return extendsTypeOption()
                .map(DecodeType.class::cast)
                .orElseGet(() -> this.baseTypeOption().get());
    }

    default EnumTypeOrTypeMeasure eitherExtendsOrBaseType() {
        Optional<EnumType> enumType = extendsTypeOption();
        return enumType.isPresent()
                ? new EnumTypeOrTypeMeasure(enumType.get())
                : new EnumTypeOrTypeMeasure(extendsOrBaseTypeProxy().typeMeasure().get()); // must not be null
    }

    default String systemName() {
        return "EnumType@" + hashCode();
    }

    default Set<EnumConstant> allConstants() {
        Set<EnumConstant> result = new HashSet<>(constants());
        result.addAll(extendsTypeOption().map(EnumType::allConstants).orElseGet(Collections::emptySet));
        return result;
    }

    @Override
    default <T> T accept(ContainerVisitor<T> visitor) {
        return visitor.visit(this);
    }

    default <T> T accept(ReferenceableVisitor<T> visitor) {
        return visitor.visit(this);
    }

}
